### PR TITLE
fe2-Android Advancing on the Digital Cash UI 

### PR DESCRIPTION
**Pull request by zoemarin**
Creating some view model and Fragment

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashMain.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+import com.google.android.material.bottomnavigation.BottomNavigationView;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Supplier;
+
+import dagger.hilt.android.AndroidEntryPoint;
+
+/**
+ * Activity for the social media
```
Comment (jbsv) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashMain.java:29]: social media ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 2]
```diff
@@ -204,6 +207,20 @@ private void subscribeSocialMediaEvent() {
                 setupSocialMediaActivity();
               }
             });
+    }
+
+    private void subscribeDigitalCashEvent() {
+      // Subscribe to "open social media" event
```
Comment (jbsv) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:213]: social media ?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 3]
```diff
@@ -262,6 +279,11 @@ public void setupSocialMediaButton() {
     socialMediaButton.setOnClickListener(v -> mViewModel.openSocialMedia());
   }
 
+  public void setupDigitalCashButton() {
+    Button socialMediaButton = findViewById(R.id.tab_digital_cash);
```
Comment (jbsv) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:283]:
```suggestion
    Button button = findViewById(R.id.tab_digital_cash);
```

**Review by jbsv — CHANGES_REQUESTED** [review]

**Comment by jbsv** [comment]
Congratulation for the big work. Just one remark for next time: it's easier for everybody to break down in several PRs, like one PR for the main menu, and a PR per sub-menu, or per fragment.

File: fe2-android/app/build.gradle [hunk 4]
```diff
@@ -108,6 +108,9 @@ dependencies {
     implementation 'androidx.constraintlayout:constraintlayout:2.1.3'
     // Android others
     implementation "androidx.lifecycle:lifecycle-reactivestreams:2.4.0"
+    implementation 'androidx.navigation:navigation-fragment:2.4.1'
+    implementation 'androidx.navigation:navigation-ui:2.4.1'
+    implementation 'androidx.legacy:legacy-support-v4:1.0.0'
```
[the remark this answers is not in the document]
Comment (zoemarin) [reply fe2-android/app/build.gradle:113]: I think it came from some beginning first try for the ui

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashMain.java [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+        setCurrentFragment(R.id.fragment_digital_cash_history, DigitalCashHistoryFragment::newInstance);
+    }
+
+    public void setupDigitalCashReceiveFragment() {
+        setCurrentFragment(R.id.fragment_digital_cash_receive, DigitalCashReceiveFragment::newInstance);
+    }
+
+    public void setupDigitalCashIssueFragment() {
+        setCurrentFragment(R.id.fragment_digital_cash_issue, DigitalCashIssueFragment::newInstance);
+    }
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashMain.java:223]: Just thinking here, maybe having a more general function setupDigitalCashActionFramgent and passing the the action as a string and having a switch statement for each action could be cleaner way to proceed.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashReceiveFragment.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+/**
+ * A simple {@link Fragment} subclass.
+ * Use the {@link DigitalCashReceiveFragment#newInstance} factory method to
+ * create an instance of this fragment.
+ */
+public class DigitalCashReceiveFragment extends Fragment {
+
+    public DigitalCashReceiveFragment() {
+        // Required empty public constructor
+    }
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashReceiveFragment.java:22]: is this necessary? Doesn't Java offer a by default empty constructor implicitly if we don't have a any constructor? I would tend to remove this, but you might have a different opinion?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashViewModel.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+   *
+   * <p>Publish a MessageGeneral containing AddChirp data.
+   *
+   * @param amount int
+   * @param sender_address String
+   * @param receiver_address String
+   */
+
+  public void sendCoin(int amount , @Nullable Address sender_address, @Nullable Address receiver_address){
+    Log.d(TAG, "Sending a transaction");
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashViewModel.java:183]: Maybe include some string representation of the sender and receiver in the log

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java (hunk 3 above)
Comment (Tesa27) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:283]: Maybe even change to `digitalCashButton` or `trasactionButton `is even clearer

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java [hunk 8]
```diff
@@ -307,17 +329,28 @@ private void setupSettingsActivity() {
...
       startActivity(intent);
     }
   }
 
+      private void setupSocialMediaActivity() {
+          if (mViewModel.getLAOs().getValue() == null) {
+              Toast.makeText(getApplicationContext(), R.string.error_no_lao, Toast.LENGTH_LONG).show();
+          } else {
+              Intent intent = new Intent(this, SocialMediaActivity.class);
+              Log.d(TAG, "Trying to open social media");
```
Comment (Tesa27) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/home/HomeActivity.java:348]: Should be changed to digital cash

**Review by Tesa27 — CHANGES_REQUESTED** [review]
Looks good overall, I have some minor suggestions

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashReceiveFragment.java (hunk 6 above)
Comment (ahominal) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashReceiveFragment.java:22]: Maybe keep it for Hilt in the future?

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashViewModel.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+  private final LAORepository laoRepository;
+  private final GlobalNetworkManager networkManager;
+  private final Gson gson;
+  private final KeyManager keyManager;
+  private final CompositeDisposable disposables;
+
+  private static final String LAO_FAILURE_MESSAGE = "failed to retrieve lao";
+  private static final String PUBLISH_MESSAGE = "sending publish message";
+
+  @Inject Wallet wallet;
```
Comment (ahominal) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashViewModel.java:58]: For what use do you need the Wallet?

**Review by ahominal — CHANGES_REQUESTED** [review]

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashSendFragment.java [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+    return mDigitalCashSendFragBinding.getRoot();
+   }
+
+   private void sendTransaction(){
+      //send some coin
+     // make a toast appear
+     if (mDigitalCashViewModel.getLaoId().getValue() == null) {
+       Toast.makeText(
+               requireContext().getApplicationContext(), R.string.error_no_lao, Toast.LENGTH_LONG)
```
[the remark this answers is not in the document]
Comment (zoemarin) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/digitalcash/DigitalCashSendFragment.java:72]: I use it to debug ^^'

**Pull request closed without merging**